cmd/collect_signals/localworker/loop.go

```diff
@@ -0,0 +1,239 @@
+	// change for a given input.
+	stateFilename string
+
+	// BucketURL is a gocloud.dev URL pointing to the data bucket where data
+	// will be written.
+	bucketURL string
+
+	// RawBucketURL is a gocloud.dev URL pointing to the raw data bucket where
+	// raw data will be written.
+	rawBucketURL string
+
+	// ShardSize defines the size of each shard passed to the worker for
+	// processing.
+	shardSize int
+}
+
+// FromConfig returns a new local WorkLoop instance using the config values
+// read in.
+//
+// Apart from the usual config values for the data bucket, raw data bucket, and
+// shard size, the criticality score config values "local-url-data-file" and
+// "local-state-file" are also used.
+func FromConfig(logger *zap.Logger, w worker.Worker) (*WorkLoop, error) {
+	if !flag.Parsed() {
+		flag.Parse()
+	}
+
+	if err := config.ReadConfig(); err != nil {
+		return nil, fmt.Errorf("reading config: %w", err)
+	}
+
+	bucketURL, err := config.GetResultDataBucketURL()
+	if err != nil {
+		return nil, fmt.Errorf("config data bucket url: %w", err)
+	}
+
+	rawBucketURL, err := config.GetRawResultDataBucketURL()
+	if err != nil {
+		return nil, fmt.Errorf("config raw data bucket url: %w", err)
+	}
+	shardSize, err := config.GetShardSize()
+	if err != nil {
+		return nil, fmt.Errorf("config shard size: %w", err)
+	}
+
+	criticalityConfig, err := config.GetCriticalityValues()
+	if err != nil {
+		return nil, fmt.Errorf("criticality score config: %w", err)
+	}
+
+	inFile := criticalityConfig[configLocalURLDataFile]
+	if inFile == "" {
+		return nil, fmt.Errorf("%q config not set", configLocalURLDataFile)
+	}
+	// This file is eventually closed when the wrapping iterator has its Close
+	// method called.
+	f, err := os.Open(inFile)
+	if err != nil {
+		return nil, fmt.Errorf("os.Open %s: %w", inFile, err)
+	}
+
+	stateFile := criticalityConfig[configLocalStateFile]
+	if stateFile == "" {
+		return nil, fmt.Errorf("%q config not set", configLocalStateFile)
+	}
+
+	return &WorkLoop{
+		logger: logger,
+		w:      w,
+
+		input:         iterator.Lines(f),
+		stateFilename: stateFile,
+
+		bucketURL:    bucketURL,
+		rawBucketURL: rawBucketURL,
+		shardSize:    shardSize,
+	}, nil
+}
+
+func (l *WorkLoop) process(ctx context.Context, req *data.ScorecardBatchRequest, bucketURL string) error {
+	exists, err := resultExists(ctx, req, bucketURL)
+	if err != nil {
+		return fmt.Errorf("result exists check: %w", err)
+	}
+
+	// Sanity check - make sure we are not re-processing an already processed request.
+	if exists {
+		l.logger.With(zap.Int32("shard", req.GetShardNum())).Info("Shard already exists. Skipping.")
+		return nil
+	}
+
+	if err := l.w.Process(ctx, req, bucketURL); err != nil {
+		return fmt.Errorf("worker process: %w", err)
+	}
+	return nil
+}
+
+// Run iterates through the repositories in batches, calling the supplied
+// worker.Process method for each batch.
+func (l *WorkLoop) Run() error {
```

Comment:
This function has a lot going on. Would it be worth splitting off some of the logic to a helper function? (e.g. the fast forward/recovery part starting line 162). I'll leave it up to your judgement.

Reply:
restore logic is in a separate function now.